presto-jdbc/src/main/java/com/facebook/presto/jdbc/ConnectionProperties.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.jdbc;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.function.Function.identity;
+
+public class ConnectionProperties
+{
+    public static final ConnectionProperty<String> USER = new User();
+    public static final ConnectionProperty<String> PASSWORD = new Password();
+    public static final ConnectionProperty<Boolean> SSL = new Ssl();
+
+    private static final Set<ConnectionProperty> ALL_OF = ImmutableSet.of(
+            USER,
+            PASSWORD,
+            SSL);
+
+    private static final Map<String, ConnectionProperty> keyLookup = unmodifiableMap(allOf()
+            .stream()
+            .collect(Collectors.toMap(ConnectionProperty::getKey, identity())));
+
+    private static final Map<String, String> DEFAULTS;
+
+    static {
+        ImmutableMap.Builder<String, String> defaults = ImmutableMap.builder();
+        for (ConnectionProperty property : ALL_OF) {
+            Optional<String> value = property.getDefault();
```

Comment:
There's some weirdness going on with the generics here, please see the comment in the updated code and let me know if you want it reverted back.